elftools/elf/dynamic.py

```diff
@@ -202,33 +296,56 @@ def iter_symbols(self):
 
         symbol_size = self.elfstructs.Elf_Sym.sizeof()
 
-        # Find closest higher pointer than tab_ptr. We'll use that to mark the
-        # end of the symbol table.
-        nearest_ptr = None
-        for tag in self.iter_tags():
-            tag_ptr = tag['d_ptr']
-            if tag['d_tag'] == 'DT_SYMENT':
-                if symbol_size != tag['d_val']:
-                    # DT_SYMENT is the size of one symbol entry. It must be the
-                    # same as returned by Elf_Sym.sizeof.
-                    raise ELFError('DT_SYMENT (%d) != Elf_Sym (%d).' %
-                                   (tag['d_val'], symbol_size))
-            if (tag_ptr > tab_ptr and
-                    (nearest_ptr is None or nearest_ptr > tag_ptr)):
-                nearest_ptr = tag_ptr
-
-        if nearest_ptr is None:
-            # Use the end of segment that contains DT_SYMTAB.
-            for segment in self.elffile.iter_segments():
-                if (segment['p_vaddr'] <= tab_ptr and
-                        tab_ptr <= (segment['p_vaddr'] + segment['p_filesz'])):
-                    nearest_ptr = segment['p_vaddr'] + segment['p_filesz']
-
-        if nearest_ptr is None:
+        end_ptr = None
+
+        # Check if a DT_GNU_HASH tag exists and recover the number of symbols
+        # from the corresponding section
+        _, gnu_hash_offset = self.get_table_offset('DT_GNU_HASH')
+        if gnu_hash_offset is not None:
+            hash_section = _DynamicGNUHashSection(self.stream, gnu_hash_offset,
+                                                  self.elffile)
+            end_ptr = tab_ptr + \
+                hash_section.get_number_of_symbols() * symbol_size
+
+        # If DT_GNU_HASH did not exist, maybe we can use DT_HASH
+        if end_ptr is None:
+            _, hash_offset = self.get_table_offset('DT_HASH')
+            if hash_offset is not None:
+                hash_section = _DynamicHashSection(self.stream, hash_offset,
+                                                   self.elffile)
+                end_ptr = tab_ptr + \
+                    hash_section.get_number_of_symbols() * symbol_size
+
+        if end_ptr is None:
```

Comment:
now that we have the better mechanisms to do this, should this arm perhaps log a warning that it is using a heuristic to detect the number of symbols?

for that matter, is it legal to have a dynamically linked object without a hashtable of some sort?

Reply:
Good idea to add a warning message, I can do this. Do you think that logic should rather be in num_symbols() as it's actually about finding the number of symbols in the table?

For the second part I'm not sure, I'll dive into the binutils code tomorrow and check. My gut feeling would be that we need one but with ELF you never know...

Reply:
I think keeping it in iter_symbols is fine, considering the way the code is architected right now.

Reply:
Alright, sounds good.

About the warning message, there is no logging in pyelftools anywhere right now, so I don't think it makes sense to add it just for this case.

I glanced over the glibc code today and from the code inside dl-lookup.c (https://github.com/lattera/glibc/blob/master/elf/dl-lookup.c#L384 and following) it looks like there must be either a DT_HASH or DT_GNU_HASH table.